protocol/l2-contract-upgrades.md

```diff
@@ -0,0 +1,189 @@
+3. update the `ProxyAdmin` (This step will only be needed the first time this scheme is used)
+4. execute a call to `ProxyAdmin.upgradePredeploys()`
+    1. This function will `DELEGATECALL` the new `L2ContractsManager`'s `upgrade()` function.
+    2. For all predeploys being upgrade, `L2ContractsManager.upgrade()` will make a call to that predeploy's `upgradeTo()` function
+
+### New L2 ProxyAdmin
+
+The `ProxyAdmin` predeploy (at `0x42...18`) will be upgraded with a new implementation which:
+
+1. maintains the current interface for backwards compatibility
+2. removes code which is necessary to support the Chugsplash and Resolved Delegate proxies (Optional, can be cut from scope)
+3. gets a new `upgradePredeploys()` function (as described above)
+    1. This function is only callable by the Depositor account (`0xdeaddead…0001`)
+
+### The L2ContractsManager contract
+
+The initial version of the L2ContractsManager developed for this work will be minimal, and its `upgrade()` function must not accept any arguments in order to maintain the determinism of the upgrade bundle.
+
+In pseudocode it will do the following:
+
+```python
+# Contains network specific config values
+class L2Config:
+    l1CrossDomainMessengerConfig: L2CrossDomainMessengerConfig
+    l2StandardBridgeConfig: L2StandardBridgeConfig
+    l2ERC721BridgeConfig: L2ERC721BridgeConfig
+    # etc.
+
+def gather_config() -> (L2Config):
+	# Calls to existing predeploys to ready any network specific config values
+	# into an L2Config object.
+
+def upgrade_and_set_config:
+	# for each upgradeable L2 contract
+	# if that contract has initializer args, call `Proxy.upgradeToAndCall()`
+	# if it does not, call `Proxy.upgradeTo()`
+
+def upgrade():
+	l2Config: L2Config = gather_config()
+	upgrade_and_set_config(l2Config)
+```
+
+### Generating deterministic JSON NUT bundles
+
+The NUT bundle will be generated via a solidity script. Similar to the
+current go-defined NUTs, this script must be deterministic, and cannot depend on execution in an EVM. The implementation addresses will therefore need to be pre-computed as seen in [the go](https://github.com/ethereum-optimism/optimism/blob/2fdd97053ba8fb2bc27dcb38dcdc42ade7c8c78d/op-node/rollup/derive/ecotone_upgrade_transactions.go#L24-L25), although this will more likely be done using the CREATE2 or CREATEX deployer.
+
+A major benefit of this approach is that it enables one to easily verify that a bundle of NUTs corresponds to a given monorepo commit, by simply:
+
+1. checking out the commit
+2. building the contracts
+3. regenerating the bundle
+4. verifying that it matches
+
+The working assumption is that the NUT bundle will be stored in the superchain-registry, as a ‘neutral’ location where any client can consume it from.
+
+**Open questions:**
+
+1. At what point should a NUT bundle for a hardfork be moved into the superchain-registry, can it simply happen when the `standard-versions-*.toml` files are updated?
```

Comment:
We have decided not to move bundles into the registry, they will be kept in the monorepo, and the correct bundle can be found via the release tag.